src/tests/unittests/d3d12/CopySplit.cpp

```diff
@@ -0,0 +1,375 @@
+        }
+    }
+
+    // Check that the offset is aligned
+    void ValidateOffset(const CopySplit& copySplit) {
+        ASSERT_TRUE(Align(copySplit.offset, D3D12_TEXTURE_DATA_PLACEMENT_ALIGNMENT) == copySplit.offset);
+    }
+
+    bool RangesOverlap(uint32_t minA, uint32_t maxA, uint32_t minB, uint32_t maxB) {
+        return (minA < minB && minB >= maxA) || (minB < minA && minA >= maxB);
+    }
+
+    // Check that no pair of copy regions intersect each other
+    void ValidateDisjoint(const CopySplit& copySplit) {
+        for (uint32_t i = 0; i < copySplit.count; ++i) {
+            const CopyInfo& a = copySplit.copies[i];
+            for (uint32_t j = i + 1; j < copySplit.count; ++j) {
+                const CopyInfo& b = copySplit.copies[j];
+                bool overlapX = RangesOverlap(a.texture.x, a.texture.x + a.texture.width, b.texture.x, b.texture.x + b.texture.width);
+                bool overlapY = RangesOverlap(a.texture.y, a.texture.y + a.texture.height, b.texture.y, b.texture.y + b.texture.height);
+                bool overlapZ = RangesOverlap(a.texture.z, a.texture.z + a.texture.depth, b.texture.z, b.texture.z + b.texture.depth);
+                ASSERT_TRUE(!overlapX || !overlapY || !overlapZ);
+            }
+        }
+    }
+
+    // Check that the union of the copy regions exactly covers the texture region
+    void ValidateTextureCoverage(const TextureSpec& textureSpec, const CopySplit& copySplit) {
+        ASSERT_TRUE(copySplit.count > 0);
+
+        uint32_t minX = copySplit.copies[0].texture.x;
+        uint32_t minY = copySplit.copies[0].texture.y;
+        uint32_t minZ = copySplit.copies[0].texture.z;
+        uint32_t maxX = copySplit.copies[0].texture.x + copySplit.copies[0].texture.width;
+        uint32_t maxY = copySplit.copies[0].texture.y + copySplit.copies[0].texture.height;
+        uint32_t maxZ = copySplit.copies[0].texture.z + copySplit.copies[0].texture.depth;
+
+        for (uint32_t i = 1; i < copySplit.count; ++i) {
+            const CopyInfo::Region& texture = copySplit.copies[i].texture;
+            minX = std::min(minX, texture.x);
+            minY = std::min(minY, texture.y);
+            minZ = std::min(minZ, texture.z);
+            maxX = std::max(maxX, texture.x + texture.width);
+            maxY = std::max(maxY, texture.y + texture.height);
+            maxZ = std::max(maxZ, texture.z + texture.depth);
+        }
+
+        ASSERT_EQ(minX, textureSpec.x);
+        ASSERT_EQ(minY, textureSpec.y);
+        ASSERT_EQ(minZ, textureSpec.z);
+        ASSERT_EQ(maxX, textureSpec.x + textureSpec.width);
+        ASSERT_EQ(maxY, textureSpec.y + textureSpec.height);
+        ASSERT_EQ(maxZ, textureSpec.z + textureSpec.depth);
+    }
+
+    // Check that copies start in the buffer exactly at the buffer offset
+    void ValidateBufferOffset(const TextureSpec& textureSpec, const BufferSpec& bufferSpec, const CopySplit& copySplit) {
```

Comment:
This could be changed to check that for each copy the first pixel is correctly copied. This plus the no-overlap plus the everything in the region check, plus the check for the number of pixels, assure that the copies are correctly putting all the pixels at the correct places.